### PR TITLE
Support LSTMBlockCell in tf2onnx mapping

### PR DESCRIPTION
**Pull request by lucienwang1009**
Map LSTMBlockCell op into ONNX ops. 
The dynamic_rnn with only a LSTMBlockCell op, which should be converted to the LSTM ONNX op by a rewriter, will be handled in later PR.

### CODE REVIEW

File: tf2onnx/tfonnx.py [hunk 1]
```diff
@@ -1683,6 +1683,143 @@ def where_op(ctx, node, name, args):
...
+    ci = tanh(ci)
+    cs = ci .* i + cs_prev .* f
+    cs = clip(cs, cell_clip)
+    o = sigmoid(cs * wco + o)
+    co = tanh(cs)
+    h = co .* o
+    ```
+    """
+    nodes = []
+    x, cs_prev, h_prev, w, wci, wcf, wco, b = node.input
```
Comment (guschmue) [inline tf2onnx/tfonnx.py]: any chance that we can make this to:
https://github.com/onnx/onnx/blob/master/docs/Operators.md#LSTM
Basically the LSTMBlock cell in tensorflow is pretty fast ... composing it from multiple ops is going to be slower.

**Comment by pengwa** [comment]
so what's this PR for?  LSTMBlockCell + static_rnn() ?

**Comment by lucienwang1009** [comment]
@pengwa This PR is just for LSTMBlockCell + Wrapper or multi-layered LSTM within dynamic_rnn, which cannot be converted into a single LSTM op.

File: tf2onnx/tfonnx.py (hunk 1 above)
Comment (lucienwang1009) [reply tf2onnx/tfonnx.py]: yes, a single LSTM is far faster than Scan op with multiple ops. But this PR is for LSTMs that cannot be converted into a single LSTM op. I will create another PR to handle the single LSTM case.

File: tf2onnx/tfonnx.py [hunk 2]
```diff
@@ -1683,6 +1683,143 @@ def where_op(ctx, node, name, args):
...
+    o = make_sigmoid(cs, wco, o)
+    # co = Tanh(cs)
+    co_node = ctx.make_node("Tanh", [cs])
+    # h = co .* o
+    h_node = ctx.make_node("Mul", [co_node.output[0], o])
+
+    def replace_output(old_output, new_output):
+        ctx.replace_all_inputs(ctx.get_nodes(), old_output, new_output)
+        ctx.copy_dtype(old_output, new_output)
+        ctx.copy_shape(old_output, new_output)
```
Comment (nbcsm) [inline tf2onnx/tfonnx.py]: one more line.

File: tf2onnx/tfonnx.py [hunk 3]
```diff
@@ -1683,6 +1683,143 @@ def where_op(ctx, node, name, args):
...
+    replace_output(node.output[2], f)
+    replace_output(node.output[3], o)
+    replace_output(node.output[4], ci_node.output[0])
+    replace_output(node.output[5], co_node.output[0])
+    replace_output(node.output[6], h_node.output[0])
+
+    nodes.extend([xh_node, xh_w_node, merged_output_node, forget_bias_const,
+                  split_output_node, f_node, ci_node, ci_i_node,
+                  cs_prev_f_node, cs_node, co_node, h_node])
+    return nodes
```
Comment (nbcsm) [inline tf2onnx/tfonnx.py]: may need to merge with Peng's graph changes after the test failures are fixed.

**Comment by pengwa** [comment]
> @pengwa This PR is just for LSTMBlockCell + Wrapper or multi-layered LSTM within dynamic_rnn, which cannot be converted into a single LSTM op.

thanks for the explanation. Yeah, due to LSTM limitation for now, we have to do this way. 

Can you put the conversion function into function folder?

File: tests/test_lstmblock.py [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+# pylint: disable=missing-docstring,invalid-name,unused-argument,using-constant-test
+
+
+class LSTMBlockTests(Tf2OnnxBackendTestBase):
+    @check_opset_min_version(8, "Scan")
+    def test_single_dynamic_lstm(self):
+        units = 5
+        batch_size = 6
+        # x_val = np.array([[1., 1.]], dtype=np.float32)
```
Comment (pengwa) [inline tests/test_lstmblock.py]: remove useless comment

File: tf2onnx/tfonnx.py [hunk 5]
```diff
@@ -1616,6 +1616,144 @@ def where_op(ctx, node, name, args):
...
+        cs = cs_clip_node.output[0]
+    # o = cs * wco + o
+    o = make_sigmoid(cs, wco, o)
+    # co = Tanh(cs)
+    co_node = ctx.make_node("Tanh", [cs])
+    # h = co .* o
+    h_node = ctx.make_node("Mul", [co_node.output[0], o])
+
+    def replace_output(old_output, new_output):
+        ctx.replace_all_inputs(ctx.get_nodes(), old_output, new_output)
```
Comment (pengwa) [inline tf2onnx/tfonnx.py]: be noted, with my PR merged, ctx.get_nodes() already get all update nodes, including the one you created. Be careful there is node having input_id == output_id after replace_all_inputs.

Comment (nbcsm) [reply tf2onnx/tfonnx.py]: maybe add a warning for such case? at least for now.

**Comment by lucienwang1009** [comment]
> > @pengwa This PR is just for LSTMBlockCell + Wrapper or multi-layered LSTM within dynamic_rnn, which cannot be converted into a single LSTM op.
> 
> thanks for the explanation. Yeah, due to LSTM limitation for now, we have to do this way.
> 
> Can you put the conversion function into function folder?

Sure.

File: tf2onnx/function/lstm_block_cell.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT license.
+
```
Comment (pengwa) [inline tf2onnx/function/lstm_block_cell.py]: add a comment for why we have this mapping.

**Pull request merged**